aten/src/ATen/native/sparse/cuda/StructuredSparseLinearCUTLASS.cu

```diff
@@ -0,0 +1,590 @@
+    // the input datatype, and then on input tensors layouts.
+    // According to the input tensors datatypes and layouts,
+    // correspnding template arguments are supplied for instantiating
+    // the wrapper function.  The tile sizes template arguments are
+    // selected according to the CUTLASS profiler results, for number
+    // of runs.
+    std::tuple<Tensor, Tensor> result;
+    AT_DISPATCH_SWITCH(
+        tensor_a.scalar_type(),
+        "_structured_sparse_linear",
+        AT_DISPATCH_CASE(
+            at::ScalarType::Char,
+            [&]() {
+                using ElementInputA = int8_t;
+                using ElementInputB = int8_t;
+                using ElementOutput = int32_t;
+                using ElementAccumulator = int32_t;
+                using ElementComputeEpilogue = int32_t;
+                using ThreadblockShape = cutlass::gemm::GemmShape<128, 128, 128>;
+                using WarpShape = cutlass::gemm::GemmShape<64, 64, 128>;
+                using InstructionShape = cutlass::gemm::GemmShape<16, 8, 64>;
+                using EpilogueOp = cutlass::epilogue::thread::LinearCombination<
+                    ElementOutput,
+                    128 / cutlass::sizeof_bits<ElementOutput>::value,
+                    ElementAccumulator,
+                    ElementComputeEpilogue>;
+                if (tensor_a_row_major && !tensor_b_row_major) {
+                    result = two_four_sgemm_cutlass<
+                        ElementInputA,
+                        ElementInputB,
+                        ElementOutput,
+                        ElementAccumulator,
+                        ElementComputeEpilogue,
+                        ThreadblockShape,
+                        WarpShape,
+                        InstructionShape,
+                        EpilogueOp,
+                        cutlass::layout::RowMajor,
+                        cutlass::layout::ColumnMajor>(
+                        tensor_a,
+                        tensor_a_stride,
+                        tensor_b,
+                        tensor_b_stride,
+                        mask_or_meta);
+                    return;
+                }
+                AT_ERROR("torch._structured_sparse_linear: Combination of "
+                         "tensor_a in ",
+                         tensor_a_row_major ? "row-major" : "column_major",
+                         " layout and tensor_b in ",
+                         tensor_b_row_major ? "row-major" : "column_major",
+                         " layout is not supported");
+            })
+        AT_DISPATCH_CASE(
+            at::ScalarType::Half,
+            [&]() {
+                using ElementInputA = cutlass::half_t;
+                using ElementInputB = cutlass::half_t;
+                using ElementOutput = cutlass::half_t;
+                using ElementAccumulator = float;
+                using ElementComputeEpilogue = cutlass::half_t;
```

Comment:
nit: any reason to not use `float` here?

Reply:
No reason, I've made the change - but as this PR is closed, the change is in the follow-up PRs: [100881](https://github.com/pytorch/pytorch/pull/100881) and [101339](https://github.com/pytorch/pytorch/pull/101339).